crosshair/libimpl/builtinslib.py

```diff
@@ -1990,71 +2034,72 @@ def rfind(self, substr, start=None, end=None):
-        return SymbolicBool(z3.SuffixOf(smt_substr, self.var))
+    def rpartition(self, sep: str):
+        result = self.rsplit(sep, maxsplit=1)
+        if len(result) == 1:
+            return ("", "", self)
+        elif len(result) == 2:
+            return (result[0], sep, result[1])
 
-    def split(self, sep: Optional[str] = None, maxsplit: int = -1):
+    def rsplit(self, sep: Optional[str] = None, maxsplit: int = -1):
         if sep is None:
-            return self.__str__().split(sep=sep, maxsplit=maxsplit)
+            return self.__str__().rsplit(sep=sep, maxsplit=maxsplit)
         smt_sep = force_to_smt_sort(sep, SymbolicStr)
         if not isinstance(maxsplit, Integral):
             raise TypeError
         if maxsplit == 0:
             return [self]
-        first_occurance = SymbolicInt(z3.IndexOf(self.var, smt_sep, 0))
-        if first_occurance == -1:
+        last_occurence = self.rfind(sep)
+        if last_occurence == -1:
             return [self]
-        ret = [self[: cast(int, first_occurance)]]
         new_maxsplit = -1 if maxsplit < 0 else maxsplit - 1
-        ret.extend(self[first_occurance + 1 :].split(sep=sep, maxsplit=new_maxsplit))
+        remaining = self[: cast(int, last_occurence)]
+        ret = self[:last_occurence].rsplit(sep, new_maxsplit)
+        index_after = len(sep) + last_occurence
+        ret.append(self[index_after:])
         return ret
 
-    def rsplit(self, sep: Optional[str] = None, maxsplit: int = -1):
+    def split(self, sep: Optional[str] = None, maxsplit: int = -1):
         if sep is None:
-            return self.__str__().rsplit(sep=sep, maxsplit=maxsplit)
+            return self.__str__().split(sep=sep, maxsplit=maxsplit)
         smt_sep = force_to_smt_sort(sep, SymbolicStr)
         if not isinstance(maxsplit, Integral):
             raise TypeError
         if maxsplit == 0:
             return [self]
-        last_occurence = self.rfind(sep)
-        if last_occurence == -1:
+        first_occurance = SymbolicInt(z3.IndexOf(self.var, smt_sep, 0))
+        if first_occurance == -1:
             return [self]
+        ret = [self[: cast(int, first_occurance)]]
         new_maxsplit = -1 if maxsplit < 0 else maxsplit - 1
-        remaining = self[: cast(int, last_occurence)]
-        ret = self[:last_occurence].rsplit(sep, new_maxsplit)
-        index_after = len(sep) + last_occurence
-        ret.append(self[index_after:])
+        ret.extend(
+            self[first_occurance + len(sep) :].split(sep=sep, maxsplit=new_maxsplit)
+        )
         return ret
 
+    def startswith(self, substr):
+        smt_substr = force_to_smt_sort(substr, SymbolicStr)
+        return SymbolicBool(z3.PrefixOf(smt_substr, self.var))
+
+    def zfill(self, width):
+        if not isinstance(width, int):
+            raise TypeError
+        if self.startswith("+") or self.startswith("-"):
+            return self[0] + self[1:].zfill(width - 1)
```

Comment:
Repeated lead characters might be a gotcha here!: 
```
>>> '++12'.zfill(10)
'+000000+12'
```

Reply:
Refactored to explicitly use only one lead character. Python semantics are weird sometimes.